### PR TITLE
Migrate rustc_codegen_llvm to SessionDiagnostics

### PR DESCRIPTION
**Pull request by SLASHLogin**
WIP: Port current implementation of diagnostics to the new SessionDiagnostics.

Part of #100717

@rustbot label +A-translation

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/errors.rs
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/errors.rs:8]:
```suggestion
pub(crate) enum UnknownCTargetFeature<'a> {
    UnknownFeaturePrefix { feature: &'a str },
    UnknownFeature { feature: &'a str, rust_feature: Option<&'a str> },
}
```

**Review by davidtwco — COMMENTED** [review]
This is a great start!

**Comment by SLASHLogin** [comment]
@davidtwco Two last things that block this are those two calls to the Handler. And I've found one method that makes use of the Handler too. 
https://github.com/rust-lang/rust/blob/2227d5117d0d8fe9d897bb3a2d020e75bb3edad5/compiler/rustc_codegen_llvm/src/lib.rs#L420-L422
https://github.com/rust-lang/rust/blob/2227d5117d0d8fe9d897bb3a2d020e75bb3edad5/compiler/rustc_codegen_llvm/src/back/lto.rs#L79-L85
https://github.com/rust-lang/rust/blob/2227d5117d0d8fe9d897bb3a2d020e75bb3edad5/compiler/rustc_codegen_llvm/src/back/write.rs#L41-L48
However, all Handlers stem from this implementation of the `WriteBackendMethods` trait. 
https://github.com/rust-lang/rust/blob/2227d5117d0d8fe9d897bb3a2d020e75bb3edad5/compiler/rustc_codegen_llvm/src/lib.rs#L168-L171
Which is then defined in the rust_codegen_ssa.rs
https://github.com/rust-lang/rust/blob/2227d5117d0d8fe9d897bb3a2d020e75bb3edad5/compiler/rustc_codegen_ssa/src/traits/write.rs#L7-L10
What's the preferred solution in this situation?

**Comment by davidtwco** [comment]
> 

You might just need to skip these for now, otherwise we would have to change `SessionDiagnostic` to use a `Handler` rather than a `ParseSess` (which might be doable).

File: compiler/rustc_codegen_llvm/src/attributes.rs
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/attributes.rs:395]: Why can't you just add a `#[help]` to the type of `TargetFeatureDisableOrEnable`?

File: compiler/rustc_codegen_llvm/src/context.rs
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/context.rs:966]: `LayoutError` could implement `SessionDiagnostic` itself if we supported `SessionDiagnostic` on enums

**Review by davidtwco — COMMENTED** [review]
This looks great, left a couple comments

File: compiler/rustc_codegen_llvm/src/attributes.rs
Comment (SLASHLogin) [reply compiler/rustc_codegen_llvm/src/attributes.rs:395]: `TargetFeatureDisableOrEnable` is present in llvm_util.rs, but without the `help`. I was thinking of splitting them into separate diagnostics, and then I could just add `#[help]` as you proposed. 
https://github.com/rust-lang/rust/blob/2227d5117d0d8fe9d897bb3a2d020e75bb3edad5/compiler/rustc_codegen_llvm/src/llvm_util.rs#L497-L502

Comment (SLASHLogin) [reply compiler/rustc_codegen_llvm/src/attributes.rs:395]: Other solution could be usage of enums instead of struct here. That would boil down to the same problem as the other comment here.

**Comment by Nilstrieb** [comment]
Hi, I've seen you changed some diagnostic structs in your PR. After #103345, the way we refer to fluent messages changed. They are now in a flat namespace with the same identifier as in the fluent file. For example,  `parser::cool_thing` is now `parser_cool_thing` and `parser::suggestion` just `suggestion`.
You should [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts) to the latest master and change your fluent message references as described above. Thanks!

**Comment by SLASHLogin** [comment]
@davidtwco I've ported remaining two diagnostics and changed old ones, which were implemented manually. However, it seems like `SharedEmitter` implements `Translate` such that it always panics, and it fails that one test. 

https://github.com/rust-lang/rust/blob/e96c330af5e3baebe7a80999744e2e082b279d0a/compiler/rustc_codegen_ssa/src/back/write.rs#L1734-L1742

https://github.com/rust-lang/rust/blob/3f66cd3ce40aa58404c901c16c211fff42a3f1a1/compiler/rustc_codegen_llvm/src/back/lto.rs#L93-L96

**Comment by SLASHLogin** [comment]
I was able to change `rustc_codegen_ssa::back::write::Diagnostic` so that it takes `Vec<(DiagnosticMessage, Style)>`, just like in `rustc_errors::diagnostic::Diagnostic`. I'm also copying arguments from the latter one to the former. However, with this approach, I have to pass new `String` when recreating `rustc_errors::Diagnostic`. Additionally, I had to implement `IntoDiagnosticArg` for `DiagnosticArgValue` to be able to pass arguments back. 
With this solution I was able to pass this UI test, but I'm failing two codegen tests though. Should I commit those changes to this branch, or should I make new PR for it?

**Comment by davidtwco** [comment]
> I was able to change `rustc_codegen_ssa::back::write::Diagnostic` so that it takes `Vec<(DiagnosticMessage, Style)>`, just like in `rustc_errors::diagnostic::Diagnostic`. I'm also copying arguments from the latter one to the former. However, with this approach, I have to pass new `String` when recreating `rustc_errors::Diagnostic`. Additionally, I had to implement `IntoDiagnosticArg` for `DiagnosticArgValue` to be able to pass arguments back. With this solution I was able to pass this UI test, but I'm failing two codegen tests though. Should I commit those changes to this branch, or should I make new PR for it?

Feel free to commit them to this branch.

File: compiler/rustc_codegen_ssa/src/back/write.rs
Comment (davidtwco) [inline compiler/rustc_codegen_ssa/src/back/write.rs:1797]: We could probably make a `d.replace_args` function that takes the whole map and just wholesale replaces its internal one with it - just to avoid some more cloning.

**Comment by davidtwco** [comment]
> With this solution I was able to pass this UI test, but I'm failing two codegen tests though.

Doesn't look like this is reproducing in CI.

File: compiler/rustc_codegen_ssa/src/back/write.rs
Comment (SLASHLogin) [reply compiler/rustc_codegen_ssa/src/back/write.rs:1797]: Sounds great to me. I can make that function and use it here.

File: compiler/rustc_codegen_llvm/src/context.rs
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/context.rs:952]: `LayoutError` implements `IntoDiagnostic`, can it be used directly here? I vaguely recall suggesting we implement `IntoDiagnostic` for `Spanned<T>` where `T: IntoDiagnostic` so that a non-spanned type like `LayoutError` can be emitted with a span without introducing a distinct type combining the two.

File: compiler/rustc_codegen_llvm/src/errors.rs
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/errors.rs:128]:
```suggestion
    #[help(codegen_llvm_missing_features)]
    pub missing_features: Option<()>,
```
This is equivalent and avoids the additional type.

File: compiler/rustc_codegen_llvm/src/errors.rs
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/errors.rs:144]: It's a shame that we need to introduce the manual impl just for this line.

File: compiler/rustc_codegen_llvm/src/context.rs
Comment (SLASHLogin) [reply compiler/rustc_codegen_llvm/src/context.rs:952]: I've had no idea Spanned existed, I can use it, sure. However with 
```rs
    #[inline]
    fn handle_layout_err(&self, err: LayoutError<'tcx>, span: Span, ty: Ty<'tcx>) -> ! {
        if let LayoutError::SizeOverflow(_) = err {
            self.sess().emit_fatal(Spanned { span, node: err })
        } else {
            span_bug!(span, "failed to get layout for `{}`: {}", ty, err)
        }
    }      
```
I have to ensure that both `'a` and `'ll` lifetimes from `&'a self` and `CodegenCx<'ll, 'tcx>` must outlive `'tcx`. And I'm unsure if I can deal with it without changing the trait's implementation.

File: compiler/rustc_codegen_llvm/src/errors.rs
Comment (SLASHLogin) [reply compiler/rustc_codegen_llvm/src/errors.rs:144]: I don't like this approach too, but at this moment there's no support for lists, so I'd stick with this manual implementation. However, after support for lists is added, I'm happy to change it. 
Alternatively, we can wait for #103397 to land.

File: compiler/rustc_codegen_llvm/src/errors.rs
Comment (SLASHLogin) [reply compiler/rustc_codegen_llvm/src/errors.rs:128]: Thanks, that's how I should have done it, but I would need `TargetFeatureDisableOrEnable` to derive `Diagnostic`, but it's manually implemented because of the 'features' is list of elements and needs it until some kind of `ListFormatter` is implemented.

File: compiler/rustc_codegen_llvm/src/errors.rs
Comment (davidtwco) [reply compiler/rustc_codegen_llvm/src/errors.rs:144]: It's okay for now.

File: compiler/rustc_codegen_llvm/src/context.rs
Comment (davidtwco) [reply compiler/rustc_codegen_llvm/src/context.rs:952]: Maybe we can just use a wrapper `SpannedLayoutError` for now, which manually implements `IntoDiagnostic, and just calls `into_diagnostic` on the layout error field, and then calls `set_span` with the `Span` field?

Comment (SLASHLogin) [reply compiler/rustc_codegen_llvm/src/context.rs]: Using this wrapper would lead to the same problem as when using `Spanned`. Instead, I've changed implementation of `IntoDiagnostic` for `LayoutError` to not bind its lifetime to the `Handler`'s and returned `DiagnosticBuilder`'s lifetime. This resolved the problem of lifetimes, so I was able to use `Spanned` as you've suggested previously.

File: compiler/rustc_codegen_ssa/src/back/write.rs
Comment (davidtwco) [inline compiler/rustc_codegen_ssa/src/back/write.rs:1790]: nit: we could probably change the constructor here so we can pass the `DiagnosticMessage` in directly

File: compiler/rustc_codegen_llvm/src/back/archive.rs
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/back/archive.rs:318]: nit:
```suggestion
            .unwrap_or_else(|kind| self.sess.emit_fatal(UnknownArchiveKind { kind }));
```

**Review by davidtwco — APPROVED** [review]
LGTM, r=me after two nits (last ones, I promise!)

File: compiler/rustc_codegen_ssa/src/back/write.rs
Comment (SLASHLogin) [reply compiler/rustc_codegen_ssa/src/back/write.rs:1790]: Constructor already is defined as `fn new<M: Into<DiagnosticMessage>>(level: Level, message: M)`, however `diag.msg` is `Vec<(DiagnosticMessage, Style)>`, so here I think it would be better to have another constructor that can take this `Vec` - something like `fn new_with_messages(level: Level, messages: Vec<(DiagnosticMessage, Style)>)`

Comment (davidtwco) [reply compiler/rustc_codegen_ssa/src/back/write.rs:1790]: Yeah, I think that makes sense.

**Comment by davidtwco** [comment]
I'm afraid that you'll need to remove the merge commit and rebase - see [no merge policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy).

**Comment by SLASHLogin** [comment]
Sorry, I've used GitHub's feature and didn't check properly that it would do the merge commit.

**Comment by SLASHLogin** [comment]
r? davidtwco

**Review by davidtwco — APPROVED** [review]

**Pull request merged**
